### PR TITLE
PUBDEV-8499: add option to display logodds for binary models

### PR DESCRIPTION
**Pull request by koniecsveta**
https://h2oai.atlassian.net/browse/PUBDEV-8499

py examples:

binary_score_form = "probability":

py_age_probabilities
py_title_probabilities

binary_score_form = "logodds":

py_age_wit_logodds
py_title_with_logodds

R examples:

binary_score_form = "probability":

Rplotsibsp
Rplotpclass

binary_score_form = "logodds":

Rplotpclass_logodds
Rplotsibsp_logodds

### CODE REVIEW

File: h2o-py/h2o/explanation/_explain.py [hunk 1]
```diff
@@ -1291,6 +1294,17 @@ def ice_plot(
     if frame.type(column) == "string":
         raise ValueError("String columns are not supported!")
 
+    if model.parms['distribution']['actual_value'] != 'bernoulli' and (binary_score_form == "logodds" or binary_score_form == "probability"):
+        raise ValueError("binary_score_form option is only supported for binomial models!")
+
+    if model.parms['distribution']['actual_value'] == 'bernoulli' and binary_score_form is None:
+        binary_score_form = "probability"
+
+    if binary_score_form != "logodds" and binary_score_form != "probability" and binary_score_form is not None:
```
Comment (johnygomez) [inline h2o-py/h2o/explanation/_explain.py:1303]: nit/optional: Maybe a little more ellegant would be `if binary_score_form not in ["logodds", "probability", None]:`

File: h2o-py/h2o/explanation/_explain.py [hunk 2]
```diff
@@ -1336,11 +1355,16 @@ def ice_plot(
...
+            logodds = np.log(tmp["mean_response"] / (1 - tmp["mean_response"]))
         if is_factor:
-            plt.scatter(factor_map(tmp.get(encoded_col)), tmp["mean_response"], color="k",
+            plt.scatter(factor_map(tmp.get(encoded_col)),
+                        tmp["mean_response"] if not show_logodds else logodds, color="k",
                         label="Partial Dependence")
         else:
-            plt.plot(tmp[encoded_col], tmp["mean_response"], color="k", linestyle="dashed",
+            plt.plot(tmp[encoded_col],
+                     tmp["mean_response"] if not show_logodds else logodds, color="k",
```
Comment (johnygomez) [inline h2o-py/h2o/explanation/_explain.py:1366]: nit/style/optional: I'd have a tendency to extract this condition, to align with the DRY principle 🙂  E.g.
```python
response = tmp["mean_response"]
if show_logodds:
    response = np.log(tmp["mean_response"] / (1 - tmp["mean_response"]))
```

File: h2o-py/tests/testdir_misc/pyunit_ice_plot.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import matplotlib
+matplotlib.use("Agg")  # remove warning from python2 (missing TKinter)
+import h2o
+import matplotlib.pyplot
+from tests import pyunit_utils
+from h2o.estimators import *
+
+
+def test_display_mode():
+    import matplotlib
```
Comment (johnygomez) [inline h2o-py/tests/testdir_misc/pyunit_ice_plot.py:15]: nit: This is I guess also the unnecessary import as in the other PR.

File: h2o-r/tests/testdir_misc/runit_ice_plot.R [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        }
+    }
+}
+
+
+
+
+doSuite("Explanation Tests", makeSuite(
+ice_plot_binary_score_form
+))
```
Comment (johnygomez) [inline h2o-r/tests/testdir_misc/runit_ice_plot.R:39]: micronit: missing new line at the end

**Review by johnygomez — APPROVED** [review]
LGTM. Seems fine from the functional perspective, I'm just adding few stylistic nits.
It shouldn't however require another review round. Thanks!

File: h2o-py/h2o/explanation/_explain.py (hunk 1 above)
Comment (koniecsveta) [reply h2o-py/h2o/explanation/_explain.py]: f3437b8

File: h2o-py/h2o/explanation/_explain.py (hunk 2 above)
Comment (koniecsveta) [reply h2o-py/h2o/explanation/_explain.py]: f3437b8

File: h2o-py/tests/testdir_misc/pyunit_ice_plot.py (hunk 3 above)
Comment (koniecsveta) [reply h2o-py/tests/testdir_misc/pyunit_ice_plot.py]: f3437b8

File: h2o-r/tests/testdir_misc/runit_ice_plot.R (hunk 4 above)
Comment (koniecsveta) [reply h2o-r/tests/testdir_misc/runit_ice_plot.R]: f3437b8

File: h2o-py/h2o/explanation/_explain.py [hunk 5]
```diff
@@ -1291,6 +1294,18 @@ def ice_plot(
     if frame.type(column) == "string":
         raise ValueError("String columns are not supported!")
 
+    distribution = _get_distribution(model)
+    if distribution not in ['bernoulli', 'binomial'] and (binary_score_form == "logodds" or binary_score_form == "probability"):
```
Comment (tomasfryda) [inline h2o-py/h2o/explanation/_explain.py:1298]: We also have `quasibinomial` and `fractionalbinomial` "distributions".  And also `custom` distribution which as far as I know can be applied on binomial classification as well. 

Maybe something like `model._model_json["output"]["model_category"] == "Binomial"` would work better? @maurever do you know if that would work correctly even for the `custom` distribution with binomial response?

Comment (maurever) [reply h2o-py/h2o/explanation/_explain.py]: I thought the `model_category` is set based on response distribution, so it is definitely better to decide with it. Good point @tomasfryda .

File: h2o-py/h2o/explanation/_explain.py [hunk 6]
```diff
@@ -1235,7 +1235,8 @@ def ice_plot(
         max_levels=30,  # type: int
         figsize=(16, 9),  # type: Union[Tuple[float], List[float]]
         colormap="plasma",  # type: str
-        save_plot_path=None # type: Optional[str]
+        save_plot_path=None, # type: Optional[str]
+        binary_score_form=None # type: Literal["probability", "logodds"]
```
Comment (tomasfryda) [inline h2o-py/h2o/explanation/_explain.py:1239]: wouldn't it be better to have just the following?
```suggestion
        binary_score_form="probability" # type: Literal["probability", "logodds"]
```

I think it would make it more clear about what is the default and you could then remove one condition in the function body (line 1301-1302). Otherwise the type declaration is incorrect (should be `Optional[Literal["probability", "logodds"]]`).

File: h2o-r/h2o-package/R/explain.R [hunk 7]
```diff
@@ -2371,7 +2389,8 @@ h2o.ice_plot <- function(model,
                          newdata,
                          column,
                          target = NULL,
-                         max_levels = 30) {
+                         max_levels = 30,
+                         binary_score_form = NULL) {
```
Comment (tomasfryda) [inline h2o-r/h2o-package/R/explain.R:2393]: Similar to python...
```suggestion
                         binary_score_form = c("probability", "logodds")) {
```

File: h2o-py/h2o/explanation/_explain.py (hunk 6 above)
Comment (koniecsveta) [reply h2o-py/h2o/explanation/_explain.py:1239]: Maybe I would go for Optional, because you don't want to use binary_score_form parameter for non binomial models

File: h2o-r/h2o-package/R/explain.R [hunk 8]
```diff
@@ -2382,6 +2401,22 @@ h2o.ice_plot <- function(model,
...
+  is_binomial <- is_binomial(model)
+  if (!is_binomial & !is.null(binary_score_form))
+    stop("binary_score_form option is only supported for binomial models!")
+  if (is_binomial & is.null(binary_score_form))
+    binary_score_form = "probability"
+  if (!is.null(binary_score_form)) {
+    if (binary_score_form != "logodds" & binary_score_form != "probability") {
+      stop("Unsupported value for binary_score_form!")
+    }
+  }
```
Comment (tomasfryda) [inline h2o-r/h2o-package/R/explain.R:2413]:
```suggestion
  if (is_binomial)
    binary_score_form <- match.arg(binary_score_form)
  }
```

File: h2o-py/h2o/explanation/_explain.py (hunk 6 above)
Comment (koniecsveta) [reply h2o-py/h2o/explanation/_explain.py]: 1f6c8de

File: h2o-r/h2o-package/R/explain.R [hunk 9]
```diff
@@ -2401,6 +2401,8 @@ h2o.ice_plot <- function(model,
   if (h2o.getTypes(newdata[[column]])[[1]] == "string")
     stop("String columns are not supported by h2o.ice_plot.")
 
+  models_info <- .process_models_or_automl(model, newdata, require_single_model = TRUE)
+
   is_binomial <- is_binomial(model)
```
Comment (tomasfryda) [inline h2o-r/h2o-package/R/explain.R:2406]: I would suggest using the following instead. The `.process_models_or_automl` does some validation but most importantly enables memoizing/caching the models on R side.

This is not important for a use-case when a user uses only one function but if they use the whole `h2o.explain` we want to limit the number of calls to `h2o.getModel` as it is very expensive in R (no lazy properties). So the `h2o.explain` uses reference class (called `models_info`) to keep track of what models were previously "materialized" on R side so we don't have to do it again BUT it also requires the functions to be able to get this object instead of the model/automl object and work with it properly. 
```r
is_binomial <- is_binomial(models_info$get_model(models_info$model))
```

Comment (koniecsveta) [reply h2o-r/h2o-package/R/explain.R]: thanks, added in c185cc1

File: h2o-py/tests/testdir_misc/pyunit_ice_plot.py [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+                break
+        else:
+            cols_to_test.append(col)
+
+    gbm = H2OGradientBoostingEstimator(seed=1234, model_id="my_awesome_model")
+    gbm.train(y=y, training_frame=train)
+
+    assert isinstance(gbm.ice_plot(train, 'title', binary_score_form="logodds").figure(), matplotlib.pyplot.Figure)
+    assert isinstance(gbm.ice_plot(train, 'age').figure(), matplotlib.pyplot.Figure)
+    matplotlib.pyplot.close("all")
```
Comment (maurever) [inline h2o-py/tests/testdir_misc/pyunit_ice_plot.py:32]: I would also test the incorrect inputs (regression model, unknown string in `binary_score_form`), to be sure everything works correctly.

**Review by maurever — APPROVED** [review]
Small improvement suggestions, otherwise LGTM.

File: h2o-py/h2o/explanation/_explain.py [hunk 11]
```diff
@@ -1291,6 +1294,18 @@ def ice_plot(
     if frame.type(column) == "string":
         raise ValueError("String columns are not supported!")
 
+    is_binomial = _is_binomial(model)
+    if (not is_binomial) and (binary_score_form == "logodds" or binary_score_form == "probability"):
```
[the remark this answers is not in the document]
Comment (koniecsveta) [reply h2o-py/h2o/explanation/_explain.py]: [b860661](https://github.com/h2oai/h2o-3/pull/6034/commits/b8606617b883959d375cf405435bd6898b2211b4)

File: h2o-py/tests/testdir_misc/pyunit_ice_plot.py (hunk 10 above)
Comment (koniecsveta) [reply h2o-py/tests/testdir_misc/pyunit_ice_plot.py:32]: [b860661](https://github.com/h2oai/h2o-3/pull/6034/commits/b8606617b883959d375cf405435bd6898b2211b4)

File: h2o-r/h2o-package/R/explain.R (hunk 7 above)
Comment (ledell) [reply h2o-r/h2o-package/R/explain.R:2393]: I agree with this suggested change by @tomasfryda.

**Review by ledell — CHANGES_REQUESTED** [review]
I would recommend changing the param name from `binary_score_form` to `binary_score_format` or `binary_prediction_scale`.

@koniecsveta @tomasfryda Question -- is this really a "probability" or just the "predicted value" of the response?  If I am not mistaken, it won't actually be a probability unless we calibrate the predicted value output using the [`calibrate_model`](https://docs.h2o.ai/h2o/latest-stable/h2o-docs/data-science/algo-params/calibrate_model.html) option?

If so, then I might suggest changing `"probability"` to `"predicted_value"`, `"response"` or maybe `"raw"`.

File: h2o-r/h2o-package/R/explain.R (hunk 7 above)
Comment (koniecsveta) [reply h2o-r/h2o-package/R/explain.R]: [b127e35](https://github.com/h2oai/h2o-3/pull/6034/commits/b127e35e1145c6725d4c8527ee6d2e8397c35e03)

File: h2o-py/h2o/explanation/_explain.py (hunk 6 above)
Comment (koniecsveta) [reply h2o-py/h2o/explanation/_explain.py]: [b127e35](https://github.com/h2oai/h2o-3/pull/6034/commits/b127e35e1145c6725d4c8527ee6d2e8397c35e03)

(changed to not optional after Erin's review)

File: h2o-r/h2o-package/R/explain.R (hunk 8 above)
Comment (koniecsveta) [reply h2o-r/h2o-package/R/explain.R]: [b127e35](https://github.com/h2oai/h2o-3/pull/6034/commits/b127e35e1145c6725d4c8527ee6d2e8397c35e03)

**Comment by koniecsveta** [comment]
> I would recommend changing the param name from `binary_score_form` to `binary_score_format` or `binary_prediction_scale`.
> 
> @koniecsveta @tomasfryda Question -- is this really a "probability" or just the "predicted value" of the response? If I am not mistaken, it won't actually be a probability unless we calibrate the predicted value output using the [`calibrate_model`](https://docs.h2o.ai/h2o/latest-stable/h2o-docs/data-science/algo-params/calibrate_model.html) option?
> 
> If so, then I might suggest changing `"probability"` to `"predicted_value"`, `"response"` or maybe `"raw"`.

it's not the calibrated value. I have renamed the option to "response" in [b127e35](https://github.com/h2oai/h2o-3/pull/6034/commits/b127e35e1145c6725d4c8527ee6d2e8397c35e03)

**Comment by ledell** [comment]
@koniecsveta I thought a bit more about using the term "score" to refer to the predicted value / outputs of the model... Usually in ML the term "score" refers to a scoring metric like "RMSE" for an entire model (rather than referring to the scale of the individual predictions).  At least, it's heavily used this way in [sklearn]https://scikit-learn.org/stable/search.html?q=score9).  To me, "binary score format" makes me think of the score of the model, not the scale of the predictions.  I'm curious to hear others' thoughts?

So I'd recommend changing `binary_score_format` to `binary_response_scale` or open to other ideas?

**Comment by koniecsveta** [comment]
> @koniecsveta I am ok with `binary_response_scale`. 👍

@ledell renamed in [fc0c02b](https://github.com/h2oai/h2o-3/pull/6034/commits/fc0c02bc5be61d5a24ab22f45b57ae415258b5f9)

**Comment by ledell** [comment]
@koniecsveta This looks good!  (i just resolved the conflicts so it should be ready to merge)... @tomasfryda what do you think?

**Review by ledell — APPROVED** [review]

**Comment by ledell** [comment]
@koniecsveta Sorry I made a mistake on fixing the conflicts!
